domain/annotation/state/state.go

```diff
@@ -0,0 +1,292 @@
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	kind := ID.Kind
+
+	// Prepare query for getting the UUID of ID
+	// No need if kind is model, because we keep annotations per model in the DB.
+	var kindQueryStmt *sqlair.Statement
+	var kindQuery string
+	var kindQueryParam sqlair.M
+
+	if kind != annotations.KindModel {
+		kindQuery, kindQueryParam, err = uuidQueryForID(ID)
+		if err != nil {
+			return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+		}
+		kindQueryStmt, err = sqlair.Prepare(kindQuery, sqlair.M{})
+		if err != nil {
+			return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+		}
+	}
+
+	// Prepare query for getting the annotations of ID
+	var getAnnotationsQuery string
+	var getAnnotationsStmt *sqlair.Statement
+
+	if kind == annotations.KindModel {
+		getAnnotationsQuery = `SELECT (key, value) AS (&Annotation.*) from annotation_model`
+	} else {
+		getAnnotationsQuery = fmt.Sprintf(`
+SELECT (key, value) AS (&Annotation.*)
+FROM annotation_%[1]s
+WHERE %[1]s_uuid = $M.uuid
+`, kind)
+	}
+
+	getAnnotationsStmt, err = sqlair.Prepare(getAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+	}
+
+	// Running transactions for getting annotations
+	var annotationsResults []Annotation
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// If it's for a model, go ahead and run the query (no uuid needed)
+		if kind == annotations.KindModel {
+			return tx.Query(ctx, getAnnotationsStmt).GetAll(&annotationsResults)
+		}
+		// Looking up the UUID for ID
+		result := sqlair.M{}
+		err = tx.Query(ctx, kindQueryStmt, kindQueryParam).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "looking up UUID for ID: %s", ID.Name)
+		}
+
+		if len(result) == 0 {
+			return errors.Annotatef(err, "unable to find UUID for ID: %q", ID.Name)
+		}
+		uuid := result["uuid"].(string)
+
+		// Querying for annotations
+		return tx.Query(ctx, getAnnotationsStmt, sqlair.M{
+			"uuid": uuid,
+		}).GetAll(&annotationsResults)
+	})
+
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// No errors, we return empty map if no annotation is found
+			return nil, nil
+		}
+		return nil, errors.Annotatef(err, "loading annotations for ID: %q", ID.Name)
+	}
+
+	annotations := make(map[string]string, len(annotationsResults))
+
+	for _, result := range annotationsResults {
+		annotations[result.Key] = result.Value
+	}
+
+	return annotations, errors.Trace(err)
+}
+
+// SetAnnotations associates key/value annotation pairs with a given ID.
+// If annotation already exists for the given ID, then it will be updated with
+// the given value. Setting a key's value to "" will remove the key from the annotations map
+// (functionally unsetting the key).
+func (st *State) SetAnnotations(ctx context.Context, ID annotations.ID,
```

Comment:
This method is doing way too much. Split it into different methods (set annotations for model, and set annotations for others).